### PR TITLE
Refactor to allow for keeping track of different key types

### PR DESCRIPTION
**Pull request by trahxam**
Changes to keep track of slices of different input types so we can run different tasks on them. Also preparing TrackML code to be more generic so we can re-use it for ITK trainings.

### CODE REVIEW

File: src/hepattn/models/maskformer.py [hunk 1]
```diff
@@ -1,74 +1,193 @@
...
-
-        assert len(self.init_nets) == 1
-        embed_x = next(iter(embed_x.values()))
-        input_pad_mask = next(iter(mask.values()))
-
-        # encoder/decoder
+            self,
+            input_nets: nn.ModuleList,
+            encoder: nn.Module,
+            decoder_layer_config: dict,
```
Comment (samvanstroud) [inline src/hepattn/models/maskformer.py:13]: Any reason to remove the encapsulation of the `Decoder` as a separate class? I feel like it makes the high level flow easier to parse.

File: src/hepattn/models/posenc.py [hunk 2]
```diff
@@ -61,23 +61,32 @@ def pos_enc(xs, dim, alpha=1000):
...
 class PositionEncoder(nn.Module):
-    def __init__(self, variables: list[str], dim: int, alpha=1000):
+    def __init__(self, feature: str, fields: list[str], sym_fields: list[str], dim: int, alpha=1000):
         """Positional encoder.
 
         Parameters
         ----------
-        variables : list[str]
-            List of variables to apply the positional encoding to.
+        feature : str
```
Comment (samvanstroud) [inline src/hepattn/models/posenc.py:69]: `feature` here is overladed, maybe `input_type` or `constituent_type` is more explicit

File: src/hepattn/models/task.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        losses = {}
+        target = targets[self.target_track + "_valid"].float()
+        weight = target + self.null_weight * (1 - target)
+        # Calculate the loss from each specified loss function.
+        for loss_fn, loss_weight in self.losses.items():
+            losses[loss_fn] = loss_weight * loss_fns[loss_fn](outputs[self.output_track + "_logit"], target, mask=None, weight=weight)
+        return losses
+
+
+class TrackHitValidTask(nn.Module):
```
Comment (samvanstroud) [inline src/hepattn/models/task.py:152]: Maybe `MaskPredictionTask`?

File: src/hepattn/experiments/tracking/tracker.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    def __init__(
+            self,
+            model: nn.Module,
+            lrs_config: dict,
+            optimizer: str = "AdamW",
+            mtl: bool = False,
+        ):
+        super().__init__(model, lrs_config, optimizer, mtl)
+
+    def log_compound_metrics(self, preds, targets, stage):
```
Comment (samvanstroud) [inline src/hepattn/experiments/tracking/tracker.py:17]: Wondering if metrics would be better suited for callbacks? Low priority in any case.

File: src/hepattn/models/maskformer.py [hunk 5]
```diff
@@ -1,74 +1,193 @@
...
+            x[feature + "_embed"] = input_net(inputs)
+            x[feature + "_valid"] = inputs[feature + "_valid"]
+
+            slice_size = inputs[feature + "_valid"].shape[-1]
+            feature_slices[feature] = slice(slice_start, slice_start + slice_size)
+            slice_start += slice_size
+
+        # Merge the input objects and he padding mask into a single set
+        # TODO: Need to correctly account for token ordering if we have multiple input features
+        x["key_embed"] = torch.concatenate([x[feature + "_embed"] for feature in input_features], dim=-2)
```
Comment (samvanstroud) [inline src/hepattn/models/maskformer.py:54]: I might be confused but by the terminology here but by`feature` do you mean `input_type`? 

In any case for the multiple input modalities, maybe some convenience functions might help like we have in salt: https://gitlab.cern.ch/atlas-flavor-tagging-tools/algorithms/salt/-/blob/main/salt/models/task.py#L54

**Review by samvanstroud — COMMENTED** [review]
Looking very nice, I had a few comments after a first quick look

File: src/hepattn/models/maskformer.py (hunk 1 above)
Comment (trahxam) [reply src/hepattn/models/maskformer.py:13]: Was done as otherwise the building of the outputs is then obfuscated inside the decoder, IMO it is nice being able to see the output building / cost building / loss pass all in one place. Also, a bunch of things e.g. the tasks, input features list, input features slicing, would have to be passed to the decoder. Open to encapsulating though if you think it hurts the readability that much (previously I wasn't using attention masks so was probably easier to read then).

File: src/hepattn/models/posenc.py (hunk 2 above)
Comment (trahxam) [reply src/hepattn/models/posenc.py:69]: Yeah I agree

File: src/hepattn/models/task.py (hunk 3 above)
Comment (trahxam) [reply src/hepattn/models/task.py:152]: Sounds good

File: src/hepattn/experiments/tracking/tracker.py (hunk 4 above)
Comment (trahxam) [reply src/hepattn/experiments/tracking/tracker.py:17]: Yeah this is probably better, especially since then metrics based on the outputs from multiple tasks can be handled in the same way, which then removes the need for us having to implement things like `log_compound_metrics` in `tracker.py`

File: src/hepattn/models/maskformer.py (hunk 5 above)
Comment (trahxam) [reply src/hepattn/models/maskformer.py:54]: I thought about doing it this way but decided against it as otherwise we have to repeat the boilerplate of getting the padding mask and masking etc everywhere whenever we want to use a specific input type in a task, although maybe it's worth doing it the other way for the sake of making `MaskFormer` cleaner

File: src/hepattn/experiments/trackml/hit_filter.py [hunk 6]
```diff
@@ -1,176 +0,0 @@
...
-        tgt = labels["hit"]["tgt_pid"].bool()
-        pred_true = pred.sigmoid() > 0.1
-        self.log(f"{stage}/nh_total_pre", float(pred.shape[1]), **kwargs)
-        self.log(f"{stage}/nh_total_post", float(pred_true.sum()), **kwargs)
-        self.log(f"{stage}/nh_pred_true", pred_true.float().sum(), **kwargs)
-        self.log(f"{stage}/nh_pred_false", (~pred_true).float().sum(), **kwargs)
-        self.log(f"{stage}/nh_valid_pre", tgt.float().sum(), **kwargs)
-        self.log(f"{stage}/nh_valid_post", (pred_true & tgt).float().sum(), **kwargs)
-        self.log(f"{stage}/nh_noise_pre", (~tgt).float().sum(), **kwargs)
-        self.log(f"{stage}/nh_noise_post", (pred_true & ~tgt).float().sum(), **kwargs)
```
Comment (samvanstroud) [inline src/hepattn/experiments/trackml/hit_filter.py:116]: I guess you plan to add this class back but here's an example of the logging

File: src/hepattn/experiments/trackml/model.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        pred_num_tracks = pred_valid.sum(-1)
+        true_num_tracks = true_valid.sum(-1)
+
+        pred_num_hits_per_track = pred_hit_masks.sum(-1).float()[pred_valid].mean()
+        true_num_hits_per_track = true_hit_masks.sum(-1).float()[true_valid].mean()
+
+        pred_num_hits_per_track_all = pred_hit_masks.sum(-1).float().mean()
+        true_num_hits_per_track_all = true_hit_masks.sum(-1).float().mean()
+
+        self.log(f"{stage}_pred_num_hits_per_track", torch.mean(pred_num_hits_per_track.float()))
```
Comment (samvanstroud) [inline src/hepattn/experiments/trackml/model.py:65]: I found out a nice feature of comet that if you use `/` in a metric name it will automatically create groups in the UI. So instead of `train_mymetric` you have `train/mymetric` and `validate/mymetric`

File: src/hepattn/experiments/trackml/prep.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
+from argparse import ArgumentParser
+from pathlib import Path
+import pandas as pd
+
+import hepattn.experiments.trackml.cluster_features as cluster_features
+
+
+# A script for preprocessing TrackML CSV files into parquet binary files
```
Comment (samvanstroud) [inline src/hepattn/experiments/trackml/prep.py:8]: Does this spead up the trackML training significantly? wondering if I should preprocess

File: src/hepattn/experiments/trackml/submit_training_hypatia.sh
Comment (samvanstroud) [inline src/hepattn/experiments/trackml/submit_training_hypatia.sh:1]: not sure why the name change here. also ideally we would abstract out the paths or duplicate the scripts

File: src/hepattn/models/input.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        Returns
+        -------
+        x : Tensor
+            Tensor containing an embedding of the concatenated input features.
+        """
+        # Some input fields will be a vector, i.e. have shape (batch, keys, D) where D > 1
+        # But must will be scalars, i.e. (batch, keys), so for these we reshape them to (batch, keys, 1)
+        # After this we can then concatenate everything together
+
+        x = self.net(concat_tensors([inputs[f"{self.input_name}_{field}"] for field in self.fields]))
```
Comment (samvanstroud) [inline src/hepattn/models/input.py:58]: Do I understand correctly here that the format for the data that gets passed into the model has changed? Now you select the relevant variables inside the model here, rather than passing in a tensor which has the features already concatted?

I think I prefer passing in a `dict[input_type: input_tensor]`, it's what we do in salt, and means that the concats can happen on the workers, less data is copied to the GPU, and there is less chance of the model accidentally using a variable it's not supposed to

File: src/hepattn/models/maskformer.py (hunk 1 above)
Comment (samvanstroud) [reply src/hepattn/models/maskformer.py:15]: ok let's leave as is for now

File: src/hepattn/models/matcher.py [hunk 10]
```diff
@@ -1,244 +1,112 @@
...
-    loss = torch.einsum("bnc,bmc->bnm", focal_pos, targets) + torch.einsum("bnc,bmc->bnm", focal_neg, (1 - targets))
-    return loss / inputs.shape[2]
+        return pred_idxs
 
+    @torch.no_grad()
+    def forward(self, costs):
+        # Cost matrix dimensions are batch, pred, true
+        # Have to detach and move the tensor from GPU to CPU then convert to numpy so we can use SciPy matcher
+        device = costs.device
+        costs = costs.detach().to(torch.float32).cpu().numpy()
```
Comment (samvanstroud) [inline src/hepattn/models/matcher.py:73]: I wonder if float16 is enough for the costs. Something to add to the list: https://github.com/samvanstroud/hepattn/issues/2

File: src/hepattn/models/transformer.py [hunk 11]
```diff
@@ -200,11 +209,13 @@ def __init__(
...
-        # Concatenate dictionary values
-        if isinstance(x, dict):
-            x = torch.cat(list(x.values()), dim=1)
-
+    def forward(self, x: Tensor, x_sort_value: Tensor | None = None, **kwargs) -> Tensor:
+        # If value to sort on is provided, use it to sort the tokens
+        # We don't need to use the stable sort assuming that the sort values are unique
+        if x_sort_value is not None:
+            x_sort_idx = torch.argsort(x_sort_value, axis=-1)
+            x = torch.gather(x, -2, x_sort_idx.unsqueeze(-1).expand_as(x))
```
Comment (samvanstroud) [inline src/hepattn/models/transformer.py:217]: This seems a bit intense for a sort? Also is there any reason this can't happen in the dataloader as before?

File: src/hepattn/models/wrapper.py [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+        for task in self.model.tasks:
+            # Check that the task actually has some metrics to log
+            if not hasattr(task, "metrics"): continue
+
+            # Just log the predictions from the final layer for now
+            task_metrics = task.metrics(preds["final"][task.name], targets)
+
+            # If the task returned a non-empty metrics dict, log it
+            if task_metrics:
+                self.log_dict({f"{stage}_{task.name}_{k}": v for k, v in task_metrics.items()})
```
Comment (samvanstroud) [inline src/hepattn/models/wrapper.py:53]:
```suggestion
                self.log_dict({f"{stage}/{task.name}_{k}": v for k, v in task_metrics.items()})
```
and similar elsewhere

File: src/hepattn/utils/awkward_hdf.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+
+
+
+
+    
+
+
+
+    print(reconstituted)
```
Comment (samvanstroud) [inline src/hepattn/utils/awkward_hdf.py:35]: Think we can remove this

File: src/hepattn/models/transformer.py (hunk 11 above)
Comment (trahxam) [reply src/hepattn/models/transformer.py:217]: I think it's much simpler that it happens before the encode stage and then is undone afterwards. If we have multiple input types (like for pflow or itk) then we would have to keep track of the ordering that each input type has in the concatenated set everywhere, which would be a bit of a nightmare

File: src/hepattn/models/matcher.py (hunk 10 above)
Comment (trahxam) [reply src/hepattn/models/matcher.py:73]: It probably is my brain just defaulted to 32. Only thing we need to be careful is that the invalid pad value used in the tasks is less than the float16 max, with enough space for multiple tasks to add that invalid pad value

File: src/hepattn/experiments/trackml/model.py (hunk 7 above)
Comment (trahxam) [reply src/hepattn/experiments/trackml/model.py:65]: Yeah we should deffo use this

File: src/hepattn/experiments/trackml/prep.py (hunk 8 above)
Comment (samvanstroud) [reply src/hepattn/experiments/trackml/prep.py:8]: Is the previous trackML dataloader still supported? I'd probably keep it if not just so my existing workflow doesn't berak :)

File: src/hepattn/models/input.py (hunk 9 above)
Comment (samvanstroud) [reply src/hepattn/models/input.py:58]: All features will always be ordered in a consistent way right? So I don't see why they can't be packed into a tensor before being shipped to the model. Is it because you lose the ability to dict-like access different features by name? (And that is useful for e.g. applying sorting inside the model?)

If we stick with this route I think we should consider a nested dict, so `data[input_type][variable_name]` rather than using `_` to separately `input_type` and `variable_name`.

File: src/hepattn/experiments/trackml/submit_training_hypatia.sh
Comment (trahxam) [reply src/hepattn/experiments/trackml/submit_training_hypatia.sh:1]: The name change is since there are multiple batch scripts (such as preprocessing, and download also for ITk), each of which have different job specs. I agree for the paths of course - what is the best / accepted way to do this?

File: src/hepattn/models/input.py (hunk 9 above)
Comment (trahxam) [reply src/hepattn/models/input.py:58]: I just remembered another reason for doing this - sometimes it is also convenient to pass around fields that are used for evaluation/plotting/eval writing, that you might not want to use in the training. If we are really squeezed for memory (shouldn't be an issue since unused fields wouldn't get gradients attached and 1M keys would be a few MB per field), we can always explicitly move them off the device in the dataloader.

File: src/hepattn/experiments/trackml/prep.py (hunk 8 above)
Comment (trahxam) [reply src/hepattn/experiments/trackml/prep.py:8]: The previous one isn't supported - does it change the existing workflow that much apart from changing the paths? (I can move a prepped sample onto lustre etc if that helps) (I'm assuming you mean just accepting CSV format and not the whole old dataloader)

File: src/hepattn/utils/cli.py [hunk 14]
```diff
@@ -50,7 +50,7 @@ def add_arguments_to_parser(self, parser) -> None:
         parser.add_argument("--name", default="hepformer", help="Name for this training run.")
         parser.add_argument("--compile", action="store_true", help="torch.compile.")
         parser.link_arguments("name", "trainer.logger.init_args.experiment_name")
-        parser.link_arguments("name", "model.name")
+        # parser.link_arguments("name", "model.name")
```
[the remark this answers is not in the document]
Comment (trahxam) [reply src/hepattn/utils/cli.py]: It was causing a crash due to me missing out model names - fixed now.

File: src/hepattn/experiments/trackml/prep.py (hunk 8 above)
Comment (samvanstroud) [reply src/hepattn/experiments/trackml/prep.py:8]: Yeah no worries I should just switch to using parquet

**Review by samvanstroud — COMMENTED** [review]
Looking pretty awesome, some more minor comments. I'm keen to get this merged to I can start using it so happy if we address them later?

File: src/hepattn/models/hitfilter.py [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+        x = {}
+
+        # Embed the input objects
+        for input_net in self.input_nets:
+            input_name = input_net.input_name
+            x[input_name + "_embed"] = input_net(inputs)
+            x[input_name + "_valid"] = inputs[input_name + "_valid"]
+
+            x[f"key_is_{input_name}"] = torch.cat([
+                torch.full((inputs[i + "_valid"].shape[-1],), i == input_name)
```
[the remark this answers is not in the document]
Comment (trahxam) [reply src/hepattn/models/hitfilter.py:34]: Good catch the device should be specified. 

We could also add a check to skip this entirely in the case where we have just one key/input type, but the extra boilerplate might not be worth the small performance gain (after adding the correct device flag).

File: src/hepattn/models/hitfilter.py [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+            x[input_name + "_valid"] = inputs[input_name + "_valid"]
+
+            x[f"key_is_{input_name}"] = torch.cat([
+                torch.full((inputs[i + "_valid"].shape[-1],), i == input_name) 
+                for i in input_names
+                ], dim=-1)
+
+        # Merge the input objects and he padding mask into a single set
+        x["key_embed"] = torch.concatenate([x[input_name + "_embed"] for input_name in input_names], dim=-2)
+        x["key_valid"] = torch.concatenate([x[input_name + "_valid"] for input_name in input_names], dim=-1)
```
[the remark this answers is not in the document]
Comment (trahxam) [reply src/hepattn/models/hitfilter.py:40]: To be consistent with maskformer.py. I agree it might be best to not include references to internally used things like this in the dict which is passed to the tasks etc. In which case we should also update maskformer.py similarly.

File: src/hepattn/callbacks/prediction_writer.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+
+    def write_layer_task_items(self, sample_group, item_name, items):
+        items_group = sample_group.create_group(item_name)
+        # This will write out a dict of items that has the structure
+        # sample/item/layer/task/value, e.g.
+        # event_name/preds/final/track_regression/track_phi
+        for layer_name, layer_items in items.items():
+            # Only write items fow the specified layers
+            if layer_name not in self.write_layers:
+                continue
```
[the remark this answers is not in the document]
Comment (trahxam) [reply src/hepattn/callbacks/prediction_writer.py:92]: Nah, more just (a) looking ahead in case at some point someone wants to do an offline analysis/detailed on how stuff changes as it passes though the model and (b) it is clean if the structure of written out data is the same as the layer/task/item structure that is used throughout the model.

File: src/hepattn/callbacks/prediction_writer.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
+import numpy as np
+import torch
+import h5py
+
+from lightning import Callback, LightningModule, Trainer
+from collections import defaultdict
+from pathlib import Path
+
+
+class TestEvalWriter(Callback):
```
[the remark this answers is not in the document]
Comment (trahxam) [reply src/hepattn/callbacks/prediction_writer.py:10]: Yeah good point.
